src/ServiceControl/Monitoring/EndpointInstanceMonitor.cs

```diff
@@ -0,0 +1,122 @@
+namespace ServiceControl.Monitoring
+{
+    using System;
+    using ServiceControl.CompositeViews.Endpoints;
+    using ServiceControl.Contracts.HeartbeatMonitoring;
+    using ServiceControl.Contracts.Operations;
+    using ServiceControl.EndpointControl.Contracts;
+    using ServiceControl.HeartbeatMonitoring;
+    using ServiceControl.Infrastructure.DomainEvents;
+
+    class EndpointInstanceMonitor
+    {
+        public EndpointInstanceId Id { get; }
+        private DateTime? lastSeen;
+        private HeartbeatStatus status;
+
+        public bool Monitored { get; private set; }
+
+        public EndpointInstanceMonitor(EndpointInstanceId endpointInstanceId, bool monitored)
+        {
+            Id = endpointInstanceId;
+            Monitored = monitored;
+        }
+
+        public void EnableMonitoring()
+        {
+            DomainEvents.Raise(new MonitoringEnabledForEndpoint { Endpoint = Convert(Id) });
+            Monitored = true;
+        }
+
+        public void DisableMonitoring()
+        {
+            DomainEvents.Raise(new MonitoringDisabledForEndpoint { Endpoint = Convert(Id) });
+            Monitored = false;
+        }
+
+        public void UpdateStatus(HeartbeatStatus newStatus, DateTime? latestTimestamp)
+        {
+            if (newStatus != status)
+            {
+                RaiseStateChangeEvents(newStatus, latestTimestamp);
+            }
+
+            lastSeen = latestTimestamp;
+            status = newStatus;
+        }
+
+        private void RaiseStateChangeEvents(HeartbeatStatus newStatus, DateTime? latestTimestamp)
+        {
+            if (newStatus == HeartbeatStatus.Alive)
+            {
+                if (status == HeartbeatStatus.Unknown)
+                {
+                    // NOTE: If an endpoint starts randomly sending heartbeats we monitor it by default
+                    // NOTE: This means we'll start monitoring endpoints sending heartbeats after a restart
+                    Monitored = true;
+                    DomainEvents.Raise(new HeartbeatingEndpointDetected
```

Comment:
The "don't monitor" flags are persisted but there is a major assumption built into this new heartbeats code: If you are sending me heartbeats then I should monitor you.

So, if an endpoint is sending heartbeats and the user marks it as "don't monitor", when you restart SC, we'll start monitoring the endpoint again. Additionally, if an endpoint is sending heartbeats and the user marks it as "don't monitor", when you restart the endpoint we'll start monitoring it again. 

I think this is the correct way to handle the situation. If you really don't want to monitor it, why is it sending heartbeats? 

If we always respected the "don't monitor" flag then you get the risk that a new (or renamed endpoint) happens to cache hit one of the "don't monitor" records and fails to show up.